### PR TITLE
Version/3.0.0

### PR DESCRIPTION
**Pull request by alexgaal**

### CODE REVIEW

File: src/Commands/MakeMiddlewareCommand.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    public function handle(): void
+    {
+        try {
+            $name = $this->argument('name');
+            $support = $this->option('support');
+            $overwrite = $this->option('overwrite');
+
+            $stub = $support ? 'middleware.support.stub' : 'middleware.stub';
+            $directory = $support ? 'Packages/Laravel/Middleware' : 'Middleware';
```
Comment (alexgaal) [inline src/Commands/MakeMiddlewareCommand.php:22]: Are we still following Spaties principles by this implementation? Idk honestly

File: src/Commands/MakeServiceProviderCommand.php [hunk 2]
```diff
@@ -6,27 +6,27 @@
...
     {
         try {
-            $className = $this->argument('className');
+            $name = $this->argument('name');
             $overwrite = $this->option('overwrite');
 
             beyond_copy_stub(
                 'service.provider.stub',
-                base_path() . "/src/App/Providers/{$className}.php",
+                base_path() . "/src/App/Providers/{$name}.php",
```
Comment (alexgaal) [inline src/Commands/MakeServiceProviderCommand.php:21]: Does the new resolver do not cover this as well?

File: src/Contracts/Schema.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+interface Schema
+{
+    public function namespace(): string;
+
+    public function namespacePath(): string;
+
+    public function className(): string;
+
+    public function path(string $directory): string;
```
Comment (alexgaal) [inline src/Contracts/Schema.php:13]: nullable would be better than empty string imho

File: src/Resolvers/AppNameSchemaResolver.php [hunk 4]
```diff
@@ -2,52 +2,64 @@
...
-        return $this->parts[2];
-    }
+        $action = new FetchDirectoryNamesFromPathAction();
 
-    public function getPath(string $directory): string
-    {
-        $parts = $this->parts;
+        $appName = $this->appName;
+        if (!$appName) {
+            $apps = $action->execute(base_path() . '/src/App');
```
Comment (alexgaal) [inline src/Resolvers/AppNameSchemaResolver.php:44]: `$action` could be named more readable

File: src/Resolvers/AppNameSchemaResolver.php [hunk 5]
```diff
@@ -2,52 +2,64 @@
...
+            do {
+                $appName = $this->command->anticipate('Please enter the app name:', $apps);
```
Comment (alexgaal) [inline src/Resolvers/AppNameSchemaResolver.php:46]: Remove `:`

File: src/Resolvers/AppNameSchemaResolver.php [hunk 6]
```diff
@@ -2,52 +2,64 @@
...
+            } while (!$appName);
+        }
 
-        array_splice($parts, 2, 0, $directory);
+        $moduleName = $this->moduleName;
+        if (!$moduleName) {
+            $modules = $action->execute(base_path() . '/src/App/' . $appName);
+            do {
+                $moduleName = $this->command->anticipate('Please enter the module name (in App/' . $appName . '):', $modules);
```
Comment (alexgaal) [inline src/Resolvers/AppNameSchemaResolver.php:54]: Remove `:`

File: src/Resolvers/AppNameSchemaResolver.php [hunk 7]
```diff
@@ -2,52 +2,64 @@
...
+        }
 
-        return implode('/', $parts);
+        return $appName . '/' . $moduleName;
     }
 
-
+    protected function askClassName(): string
+    {
+        return $this->className ?? $this->command->ask('Please enter the class name:');
```
Comment (alexgaal) [inline src/Resolvers/AppNameSchemaResolver.php:63]: Remove `:`

File: src/Resolvers/BaseNameSchemaResolver.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    ) {
+    }
+
+    abstract public function handle(): Schema;
+
+    abstract protected function askNamespace(): string;
+
+    protected function askClassName(): string
+    {
+        return $this->className ?? $this->command->ask('Please enter the class name:');
```
Comment (alexgaal) [inline src/Resolvers/BaseNameSchemaResolver.php:24]: Remove `:`

File: src/Resolvers/DomainNameSchemaResolver.php [hunk 9]
```diff
@@ -2,47 +2,42 @@
...
-    {
-        return $this->parts[1];
+        return new DomainSchema($namespace, $className);
     }
 
-    public function getPath(string $directory): string
+    protected function askNamespace(): string
     {
-        $parts = $this->parts;
+        $action = new FetchDirectoryNamesFromPathAction();
```
Comment (alexgaal) [inline src/Resolvers/DomainNameSchemaResolver.php:29]: could be more readable as well

File: src/Resolvers/DomainNameSchemaResolver.php [hunk 10]
```diff
@@ -2,47 +2,42 @@
...
+        $domains = $action->execute(base_path() . '/src/Domain');
 
-        array_splice($parts, 1, 0, $directory);
+        do {
+            $domainName = $this->command->anticipate('Please enter the domain name:', $domains);
```
Comment (alexgaal) [inline src/Resolvers/DomainNameSchemaResolver.php:33]: Remove `:`

File: src/Resolvers/DomainNameSchemaResolver.php [hunk 11]
```diff
@@ -2,47 +2,42 @@
...
+        } while (!$domainName);
 
-        return implode('/', $parts);
+        return $domainName;
     }
 
-
+    protected function askClassName(): string
+    {
+        return $this->className ?? $this->command->ask('Please enter the class name:');
```
Comment (alexgaal) [inline src/Resolvers/DomainNameSchemaResolver.php:41]: Remove `:`

File: src/Schema/BaseSchema.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        return str_replace('\\', '/', $this->namespace);
+    }
+
+    public function className(): string
+    {
+        return $this->className;
+    }
+
+    abstract public function path(string $directory): string;
```
Comment (alexgaal) [inline src/Schema/BaseSchema.php:30]: nullable parameter

File: src/Commands/MakeCommandCommand.php [hunk 13]
```diff
@@ -3,29 +3,32 @@
...
     protected $description = 'Make a new command';
 
     public function handle()
     {
-        $className = $this->argument('className');
+        $name = $this->argument('name');
         $command = $this->option('command');
         $overwrite = $this->option('overwrite');
 
+        $schema = (new AppNameSchemaResolver($this, $name, 'Command', 'Console'))->handle();
```
Comment (alexgaal) [inline src/Commands/MakeCommandCommand.php:20]: should be `Commands`

File: src/Commands/MakeQueryBuilderCommand.php [hunk 14]
```diff
@@ -18,28 +17,28 @@ public function handle(): void
             $name = $this->argument('name');
             $overwrite = $this->option('overwrite');
 
-            $schema = new DomainNameSchemaResolver($name);
+            $schema = (new DomainNameSchemaResolver($this, $name))->handle();
 
             beyond_copy_stub(
                 'query-builder.stub',
-                base_path() . '/src/Domain/' . $schema->getPath('QueryBuilders') . '.php',
+                $schema->path('QueryBuilders'),
```
Comment (alexgaal) [inline src/Commands/MakeQueryBuilderCommand.php:24]: I'd prefer Laravel standard: `Builder`.

File: src/Actions/CopyFileAction.php [hunk 15]
```diff
@@ -13,8 +13,6 @@ public function execute(string $srcPath, string $targetPath, bool $overwrite = f
 
         $fs->ensureDirectoryExists(
             dirname($targetPath),
-            0755,
```
[the remark this answers is not in the document]
Comment (regnerisch) [reply src/Actions/CopyFileAction.php:16]: As it is the default values, I do not know why we added those.

File: .php-cs-fixer.php [hunk 16]
```diff
@@ -0,0 +1,17 @@
+<?php
```
[the remark this answers is not in the document]
Comment (alexgaal) [reply .php-cs-fixer.php:1]: because it's Laravel standard and we'd stick to their Code Style.

File: src/Commands/MakeMiddlewareCommand.php (hunk 1 above)
Comment (regnerisch) [reply src/Commands/MakeMiddlewareCommand.php:22]: Yes. As I remember middlewares are put to `Support` and app specific middlewares are put inside `App`

File: src/Commands/MakeQueryBuilderCommand.php (hunk 14 above)
Comment (regnerisch) [reply src/Commands/MakeQueryBuilderCommand.php:24]: `Builders` as we use plural everywhere

Comment (alexgaal) [reply src/Commands/MakeQueryBuilderCommand.php:24]: I'll change.

**Review by regnerisch — APPROVED** [review]

**Pull request merged**